### PR TITLE
Move folding & visiting traits into type library

### PR DESCRIPTION
**Pull request by eggyal**
This is a rework of #107712, following feedback on that PR.

In particular, this version uses trait aliases to reduce the API churn for trait consumers.  Doing so requires a workaround for #107747 until its fix in #107803 is merged into the stage0 compiler; this workaround, which uses conditional compilation based on the `bootstrap` configuration predicate, sits in dedicated commit b409329c for ease of reversion.

The possibility of the `rustc_middle` crate retaining its own distinct versions of each folding/visiting trait, blanket-implemented on all types that implement the respective trait in the type library, was also explored: however since this would necessitate making each `rustc_middle` trait a subtrait of the respective type library trait (so that such blanket implementations can delegate their generic methods), no benefit would be gained.

r? types

### CODE REVIEW

**Comment by eggyal** [comment]
I should add that, in contrast to the previous PR, this version splits methods from the `TypeVisitable` trait into an extension trait that is retained solely in `rustc_middle`.

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -1120,6 +1120,13 @@ impl From<call::AdjustForForeignAbiError> for FnAbiError<'_> {
 
 impl<'tcx> fmt::Display for FnAbiError<'tcx> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        #[cfg(bootstrap)]
+        match self {
+            Self::Layout(err) => fmt::Display::fmt(err, f),
+            Self::AdjustForForeignAbi(err) => fmt::Display::fmt(err, f),
+        }
+
+        #[cfg(not(bootstrap))]
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/layout.rs:1129]: why is this necessary?

Comment (oli-obk) [reply compiler/rustc_middle/src/ty/layout.rs:1129]: edit: ah! due to the super trait stuff.
